lib/LWP/UserAgent.pm

```diff
@@ -1140,7 +1140,24 @@ sub proxy {
         my $url = shift;
         if (defined($url) && length($url)) {
             Carp::croak("Proxy must be specified as absolute URI; '$url' is not") unless $url =~ /^$URI::scheme_re:/;
-            Carp::croak("Bad http proxy specification '$url'") if $url =~ /^https?:/ && $url !~ m,^https?://[\w[],;
+            # Proxy host might be an IPv4, IPv6 or a domain name with optional port separated by colon
+            # Additionally, I saw sometimes a trailing slash in the proxy path
+            my $host_re = '^[\w\[\]:.-]+/?$';
+            if ( $url =~ m!^(https?://)(.*)! ) {
+                my $scheme = $1;
+                my $path = $2;
+                if ($path =~ /@/) {
+                    my ($credentials, $host) = $path =~ /^(.+)@(.*)$/;
+                    Carp::croak("Bad http proxy specification with '$url'") unless $host && $host =~ m/$host_re/;
+                    my ($user, $pass) = $credentials =~ /^(.*):(.*)$/;
+                    Carp::croak("Neither user nor password can contain ':' symbol") if $user =~ /:/;
+                    $user =~ s/([^\w])/sprintf("%%%0x", ord($1))/ge;
```

Comment:
Should we not be using [URI](https://metacpan.org/pod/URI) directly? If we don't want to cleanse that URI with URI directly, then maybe [URI::Escape::uri_escape_utf8()](https://metacpan.org/pod/URI::Escape#uri_escape_utf8(-$string-)) instead of redoing it ourselves here?

Just attempt to make a URI object instance, then grab the `->as_string` of it to have a properly dealt with URI string.

Apologies if I'm missing something, but it seems like we're adding yet another place for URI encoding rather than using the tools we already have a dependency on. If the URI module needs updating after testing this way, we should likely fix that instead, right?

Reply:
Yeah, actually you're right, we can use URI / URI::Escape modules here.
Just one moment, should I load them globally or require inside proxy subroutine?
I see that in lib/LWP/Protocol/http.pm the URI::Escape module is loaded only when needed, I guess there should be the same behavior, am I right?